Add support for ! in Conditional IDL attributes.

xeenon opened the pull request:
#### f862b9eefaacaacf7664bacad44054a4d3ff1533
<pre>
Add support for ! in Conditional IDL attributes.
<a href="https://webkit.org/b/298793">https://webkit.org/b/298793</a>

Reviewed by Darin Adler and Brian Weinstein.

This change allows `Conditional` attribute values in IDL files to be prefixed with `!`, which outputs
a negated form of `ENABLE()`. Because `!` is not a valid character in IDL attribute values, the value
must be wrapped in quotes when used. Bare conditionals continue to work as before, maintaining
compatibility with existing IDL files.

Example:

  [Conditional=&quot;!FEATURE_FLAG&quot;] readonly attribute double someName;

Cleaned up the function to use more idiomatic Perl with proper use of map.

* Source/WebCore/bindings/scripts/CodeGenerator.pm:
(GenerateConditionalStringFromAttributeValue):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(AddToIncludes):
* Source/WebCore/bindings/scripts/test/DOMWindowConstructors.idl:
* Source/WebCore/bindings/scripts/test/JS/JSDOMWindow.cpp:
(WebCore::JSDOMWindowDOMConstructor::prototypeForStructure):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
(WebCore::jsDOMWindow_TestConditionalConstructorGetter):
* Source/WebCore/bindings/scripts/test/JS/JSTestConditional.cpp: Added.
(WebCore::JSTestConditionalDOMConstructor::prototypeForStructure):
(WebCore::JSTestConditionalDOMConstructor::initializeProperties):
(WebCore::JSTestConditionalPrototype::finishCreation):
(WebCore::JSTestConditional::JSTestConditional):
(WebCore::JSTestConditional::createPrototype):
(WebCore::JSTestConditional::prototype):
(WebCore::JSTestConditional::getConstructor):
(WebCore::JSTestConditional::destroy):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
(WebCore::jsTestConditional_fooEnabledGetter):
(WebCore::jsTestConditional_fooDisabledGetter):
(WebCore::jsTestConditional_barOrBazEnabledGetter):
(WebCore::jsTestConditional_fooEnabledOrBarDisabledGetter):
(WebCore::jsTestConditional_fooOrBarBothDisabledGetter):
(WebCore::jsTestConditional_fooAndBarEnabledGetter):
(WebCore::jsTestConditional_fooAndBarDisabledGetter):
(WebCore::jsTestConditional_fooAndBarBothDisabledGetter):
(WebCore::jsTestConditional_fooOrBarAndBazGetter):
(WebCore::jsTestConditional_fooAndBarOrBazGetter):
(WebCore::jsTestConditional_fooOrBarAndBazDisabledGetter):
(WebCore::jsTestConditional_fooAndBarDisabledOrBazGetter):
(WebCore::jsTestConditional_fooDisabledOrBarAndBazGetter):
(WebCore::JSTestConditional::subspaceForImpl):
(WebCore::JSTestConditional::analyzeHeap):
(WebCore::JSTestConditionalOwner::isReachableFromOpaqueRoots):
(WebCore::JSTestConditionalOwner::finalize):
(WebCore::verifyVTable):
(WebCore::toJSNewlyCreated):
(WebCore::toJS):
(WebCore::JSTestConditional::toWrapped):
* Source/WebCore/bindings/scripts/test/JS/JSTestConditional.h: Added.
(WebCore::JSTestConditional::create):
(WebCore::JSTestConditional::createStructure):
(WebCore::JSTestConditional::subspaceFor):
(WebCore::wrapperOwner):
(WebCore::wrapperKey):
(WebCore::toJS):
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/scripts/test/SupplementalDependencies.dep:
* Source/WebCore/bindings/scripts/test/TestConditional.idl: Added.

Canonical link: <a href="https://commits.webkit.org/299931@main">https://commits.webkit.org/299931@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d16cee6436a477d6f3b65e3e256c48b900aa01e6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120778 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40472 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31125 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127183 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72854 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/cf32948c-38bf-4b91-98dd-6c9e2243d9e4) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/122654 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41170 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49049 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/91743 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/60987 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/84735abd-4856-4964-919f-4b530ccf8b8a) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123730 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32869 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/108262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72375 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/21d4e7fe-08d9-43f2-aefa-b2014db27807) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31899 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/26369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70779 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102362 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/26547 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130046 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47699 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/36219 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/100360 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48067 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/104440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100244 "Passed tests") | | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/45639 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/23680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/44368 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19166 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Checked out pull request; Reviewed by Darin Adler and Brian Weinstein; Canonicalized commit; Pushed commit to WebKit repository") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47561 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53266 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47030 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50376 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48716 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/bindings/scripts/CodeGenerator.pm:
```diff
@@ -1464,26 +1464,20 @@ sub GenerateConditionalStringFromAttributeValue
-        $expression = "($expression)" if $expression =~ / /;
-        push @parenthesized, $expression;
-    } sort keys %disjunction;
+        join(" && ", sort keys %conjunction) => 1
```
This sort, and the one below, orders everything negated before everything not negated, which is OK, I guess, but a bit arbitrary. We could instead write a sort comparator function that moves the "!" to the end of the string before comparing (something like `s/(!?)(.+)/$2$1/`), which would sort by the conditional name instead.

darinadler commented on Source/WebCore/bindings/scripts/CodeGenerator.pm:
```diff
@@ -1464,26 +1464,20 @@ sub GenerateConditionalStringFromAttributeValue
-    return (%disjunction)[0] if keys %disjunction == 1;
+    my %disjunction = map {
+        my %conjunction = map {
+            /^!(.+)$/ ? ("!ENABLE($1)" => 1) : ("ENABLE($_)" => 1)
```
I would slightly prefer something more like `s/^(!?)(.+)$/$1ENABLE($2)/` rather than using the `?:` operator.

xeenon committed:
Add support for ! in Conditional IDL attributes.
https://webkit.org/b/298793

Reviewed by Darin Adler and Brian Weinstein.

This change allows `Conditional` attribute values in IDL files to be prefixed with `!`, which outputs
a negated form of `ENABLE()`. Because `!` is not a valid character in IDL attribute values, the value
must be wrapped in quotes when used. Bare conditionals continue to work as before, maintaining
compatibility with existing IDL files.

Example:

  [Conditional="!FEATURE_FLAG"] readonly attribute double someName;

Cleaned up the function to use more idiomatic Perl with proper use of map.

* Source/WebCore/bindings/scripts/CodeGenerator.pm:
(GenerateConditionalStringFromAttributeValue):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(AddToIncludes):
* Source/WebCore/bindings/scripts/test/DOMWindowConstructors.idl:
* Source/WebCore/bindings/scripts/test/JS/JSDOMWindow.cpp:
(WebCore::JSDOMWindowDOMConstructor::prototypeForStructure):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
(WebCore::jsDOMWindow_TestConditionalConstructorGetter):
* Source/WebCore/bindings/scripts/test/JS/JSTestConditional.cpp: Added.
(WebCore::JSTestConditionalDOMConstructor::prototypeForStructure):
(WebCore::JSTestConditionalDOMConstructor::initializeProperties):
(WebCore::JSTestConditionalPrototype::finishCreation):
(WebCore::JSTestConditional::JSTestConditional):
(WebCore::JSTestConditional::createPrototype):
(WebCore::JSTestConditional::prototype):
(WebCore::JSTestConditional::getConstructor):
(WebCore::JSTestConditional::destroy):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
(WebCore::jsTestConditional_fooEnabledGetter):
(WebCore::jsTestConditional_fooDisabledGetter):
(WebCore::jsTestConditional_barOrBazEnabledGetter):
(WebCore::jsTestConditional_fooEnabledOrBarDisabledGetter):
(WebCore::jsTestConditional_fooOrBarBothDisabledGetter):
(WebCore::jsTestConditional_fooAndBarEnabledGetter):
(WebCore::jsTestConditional_fooAndBarDisabledGetter):
(WebCore::jsTestConditional_fooAndBarBothDisabledGetter):
(WebCore::jsTestConditional_fooOrBarAndBazGetter):
(WebCore::jsTestConditional_fooAndBarOrBazGetter):
(WebCore::jsTestConditional_fooOrBarAndBazDisabledGetter):
(WebCore::jsTestConditional_fooAndBarDisabledOrBazGetter):
(WebCore::jsTestConditional_fooDisabledOrBarAndBazGetter):
(WebCore::JSTestConditional::subspaceForImpl):
(WebCore::JSTestConditional::analyzeHeap):
(WebCore::JSTestConditionalOwner::isReachableFromOpaqueRoots):
(WebCore::JSTestConditionalOwner::finalize):
(WebCore::verifyVTable):
(WebCore::toJSNewlyCreated):
(WebCore::toJS):
(WebCore::JSTestConditional::toWrapped):
* Source/WebCore/bindings/scripts/test/JS/JSTestConditional.h: Added.
(WebCore::JSTestConditional::create):
(WebCore::JSTestConditional::createStructure):
(WebCore::JSTestConditional::subspaceFor):
(WebCore::wrapperOwner):
(WebCore::wrapperKey):
(WebCore::toJS):
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/scripts/test/SupplementalDependencies.dep:
* Source/WebCore/bindings/scripts/test/TestConditional.idl: Added.

Canonical link: https://commits.webkit.org/299931@main
